### ISSUE TITLE
ty fails to infer type of decorated function

### ISSUE THREAD

**Issue by randolf-scholz** [astral-sh/ty#4185]
### Summary

In the example below `decorator` is essentially an identity mapping. However, when decorating `pprint_mapping` with `@decorator`, `ty`  infers `pprint_mapping` as `(X@decorator, /) -> X@decorator` rather than `[Map](cls: type[Map]) -> type[Map]`.

`main.py`:

```py
from typing import Any, Protocol, overload, Callable as Fn, reveal_type, Concatenate
from collections.abc import Mapping

def decorator[X, Y](fn: Fn[[X], Y]) -> Fn[[X], Y]:
    return fn

@decorator
def pprint_mapping[Map: Mapping](cls: type[Map]) -> type[Map]:
    raise NotImplementedError


reveal_type(pprint_mapping)

@pprint_mapping  # E: Expected `X@decorator`, found `<class 'MyMapping'>`
class MyMapping(Mapping[str, int]): ...

reveal_type(MyMapping)
```

`ty.json`:

```json
{
    "environment": {
        "python-version": "3.14"
    },
    "rules": {
        "experimental-syntax": "ignore",
        "undefined-reveal": "ignore"
    }
}
```

[Playground](https://play.ty.dev/f159798f-acb7-4696-b9ce-1719bb2aaaa6)

**Comment by randolf-scholz** [comment]
A more non-trivial example: https://play.ty.dev/2d2413b2-8ece-4b7e-aaa3-abd7c10c070e

**Comment by carljm** [comment]
Thanks, this is #3186
